File: LeetCode/python/N-QueensII.py
class Solution(object):
    # chessBoard的下标是col,值是row
    def isSafeHelper(self,chessBoard,col,row):
        result = False
        
        for i in range(0,col):
            if(row == chessBoard[i]):
                break
            
            if(row - chessBoard[i] == col - i):
                break
                
            if(row - chessBoard[i] == i - col):
                break
            
            if(i == col - 1):
                result = True

        return result
    
    def isSafe(self,chessBoard,col,n):
        cnt = 0
        if(col == n):
            cnt = 1
        else:            
            for i in range(0,n):
                if(self.isSafeHelper(chessBoard,col,i)):
                    chessBoard[col] = i
                    cnt += self.isSafe(chessBoard,col+1,n)
                    
        return cnt

    def totalNQueens(self, n):
        """
        :type n: int
        :rtype: int
        """
        QueensCnt = 0
        
        chessBoard = []
        
        for i in range(0,n):
            chessBoard.append(-1)
        
        for i in range(0,n):
            chessBoard[0] = i #第一列
            QueensCnt += self.isSafe(chessBoard,1,n)
			
        return QueensCnt
				
				
#测试程序
s = Solution()

print(s.totalNQueens(8))

File: LeetCode/c++/N-QueensII.cpp
#include <iostream>

using namespace std;


bool isSafeHelper(int *chessBoard, int row, int col)
{
	bool result = false;

	for (int i = 0; i<col; i++)
	{
		if (chessBoard[i] == row)
			break;
		
		//对角
		if (row - chessBoard[i] == col - i)
			break;

		//斜对角
		if (row - chessBoard[i] == i - col)
			break;

		if (i == col - 1)
			result = true;
	}

	return result;
}

int isSafe(int *chessBoard, int col, int n)
{
	int cnt = 0;

	if (col == n)
	{
		cnt++;
	}
	else
	{
		for (int i = 0; i < n; i++)
		{
			if (isSafeHelper(chessBoard, i, col))
			{
				chessBoard[col] = i;
				cnt += isSafe(chessBoard, col + 1, n);
			}

		}
	}

	return cnt;
}

int totalNQueens(int n) {	
	int queenCnt = 0;

	//chessBoard[col] = row, 下标是列，数组的值是该列皇后所在的行
	int *chessBoard = new int[n];

	for (int i = 0; i<n; i++)
	{
		chessBoard[0] = i;

		queenCnt += isSafe(chessBoard, 1, n);
	}

	return queenCnt;
}int main()
{
	cout << totalNQueens(8);

	return 0;
}


